add snapshot page detailing legacy syntax (#6803)
fishtown-analytics/docs.getdbt.com: website/docs/reference/resource-configs/snapshots-jinja-legacy.md

---
title: Legacy snapshot configurations
description: Read about how to configure snapshots using legacy jinja blocks
sidebar: "Legacy configuration"
---

# Legacy snapshot configuration <Lifecycle status='legacy' />

<IntroText>

> Configure snapshots using Jinja blocks in your `.sql` files, which is available now and considered legacy syntax.

Author: ```suggestion
Use legacy SQL-based snapshot configurations with Jinja blocks in any dbt version. dbt v1.9 introduced YAML-based configs for better readability and environment awareness.
```

</IntroText>

There are situations where you want to use the legacy syntax for [snapshots](/docs/build/snapshots) in any dbt version or release track. This page details how you can use the legacy SQL-based configurations if you need to. 

In dbt v1.9, this syntax was replaced with a [YAML-based configuration](/reference/snapshot-configs#configuring-snapshots) in [dbt Cloud's "Latest" release track](/docs/dbt-versions/cloud-release-tracks). The benefits of YAML-based configurations are that the snapshots are environment aware, meaning you don't have to specify `schema` or `database`, and the syntax is more concise.

For new snapshots, we recommend using these latest YAML-based configs. If you'd like to move to the YAML-based configuration for existing snapshots, you can [migrate over](/reference/snapshot-configs#snapshot-configuration-migration).

When would you want to use the SQL-based syntax and YAML-based syntax?

- SQL-based syntax:

>   - Defined in `.sql` files within a snapshot Jinja block, typically located in your `snapshots` directory. Available in all versions.

Reviewer: Hiya @mirnawong1 

Should Jinja be capitalised or lower case? It's in lower case in your intro so just double checking for the caps use here.

Kind Regards
Natalie

Author: oops, should be capitalized! good catch!

</VersionBlock>

<VersionBlock firstVersion="1.9">

<File name='snapshots/orders_snapshot.sql'>

```sql
{% snapshot orders_snapshot %}

{{
    config(
      database='analytics',
      schema='snapshots',
      unique_key='id',

      strategy='timestamp',
      updated_at='updated_at',
    )
The following table outlines the configurations available for snapshots:

select * from {{ source('jaffle_shop', 'orders') }}

{% endsnapshot %}
```

</File>
</VersionBlock>

6. Run the `dbt snapshot` [command](/reference/commands/snapshot). For our example, a new table will be created at `analytics.snapshots.orders_snapshot`. You can change the `target_database` configuration, the `target_schema` configuration and the name of the snapshot (as defined in `{% snapshot .. %}`) will change how dbt names this table.

```
dbt snapshot
Running with dbt=1.8.0

15:07:36 | Concurrency: 8 threads (target='dev')
15:07:36 |
15:07:36 | 1 of 1 START snapshot snapshots.orders_snapshot...... [RUN]
15:07:36 | 1 of 1 OK snapshot snapshots.orders_snapshot..........[SELECT 3 in 1.82s]
15:07:36 |
15:07:36 | Finished running 1 snapshots in 0.68s.

Completed successfully

Done. PASS=2 ERROR=0 SKIP=0 TOTAL=1
```

> 7. Inspect the results by selecting from the table dbt created. After the first run, you should see the results of your query, plus the [snapshot meta fields](#snapshot-meta-fields) as described earlier.

Reviewer: Hiya @mirnawong1 

For some reason, this link didn't work for me. It didn't direct me to the relevant page. The same link above took me to the relevant page though.

Kind Regards

Author: good catch! i copied this over from the snapshots page, which was linking to another section. thank you - updated!

---
title: Legacy snapshot configurations
description: Read about legacy snapshot jinja blocks and how to migrate to the updated syntax
sidebar: Legacy snapshot configurations
---

# Legacy snapshot configuration <Lifecycle status='legacy' />

> From dbt versions 1.8 and earlier, you were able to configure [snapshots](/docs/build/snapshots) using jinja blocks in your `.sql` files. This is considered legacy syntax and was replaced with a [YAML-based configuration](/reference/snapshot-configs#configuring-snapshots) in [dbt Cloud's "Latest" release track](/docs/dbt-versions/cloud-release-tracks) and dbt v1.9 for faster and more efficient management. 

Author: available now

> From dbt versions 1.8 and earlier, you were able to configure [snapshots](/docs/build/snapshots) using jinja blocks in your `.sql` files. This is considered legacy syntax and was replaced with a [YAML-based configuration](/reference/snapshot-configs#configuring-snapshots) in [dbt Cloud's "Latest" release track](/docs/dbt-versions/cloud-release-tracks) and dbt v1.9 for faster and more efficient management. 

Author: update language so its clear its avail now as well

> From dbt versions 1.8 and earlier, you were able to configure [snapshots](/docs/build/snapshots) using jinja blocks in your `.sql` files. This is considered legacy syntax and was replaced with a [YAML-based configuration](/reference/snapshot-configs#configuring-snapshots) in [dbt Cloud's "Latest" release track](/docs/dbt-versions/cloud-release-tracks) and dbt v1.9 for faster and more efficient management. 

Author: look at a snapshot blog and link to it + summarize
- legacy syntax still supported, this explains how to use it. if interested in migrating, check out this guide. 
- we recommend using latest yaml bc ...
- outlines diff legacy/yaml - can remove this
- change using - one of the most important configs you can decide is strategies which tells dbt how to detect modified rows.
  - You can also configure snapshots using [strategies](https://docs-getdbt-com-git-legacy-snapshot-jinja-dbt-labs.vercel.app/reference/resource-configs/snapshots-jinja-legacy?strategy=check#snapshot-strategies), which define how dbt knows if a row has changed.

However, there are situations where you might still need to use the legacy syntax for snapshots in any dbt version or release track. This page details how you can use the legacy SQL-based configurations and provides a path to migrate to the more efficient YAML configuration. 

For new snapshots, we recommend using these latest YAML-based configs. If applying them to existing snapshots, you'll need to [migrate over](/reference/snapshot-configs#snapshot-configuration-migration).

The following outlines the differences between the legacy SQL-based syntax and the updated YAML-based syntax:

<Expandable alt_header="SQL-based snapshot syntax">

Legacy syntax for defining snapshots in `.sql` files within a snapshot Jinja block, typically located in your `snapshots` directory. Available in dbt v1.8 and earlier.

#### Use cases

> - Defining multiple resources in a single file, though it can significantly slow down parsing and compilation.

Author: ```suggestion

```

- Useful for existing snapshots already using this syntax.
- Suitable for performing very light transformations (but creating a separate ephemeral model for transformations is recommended for better maintainability).

#### Example

In this example, we created a snapshot in the `snapshots` directory and used the `{{ config() }}` block within the jinja block to define the snapshot configuration.

<File name='snapshots/orders_snapshot.sql'>

```sql
{% snapshot orders_snapshot %}

{{ config(
    target_database='analytics',
    target_schema='snapshots',
    unique_key='id',
    strategy='timestamp',
    updated_at='updated_at'
) }}

select * from {{ source('jaffle_shop', 'orders') }}

{% endsnapshot %}
```
</File>
</Expandable>

<Expandable alt_header="YAML-based snapshot syntax">

> Updated syntax for defining snapshot configurations in YAML files. Found in `snapshots.yml`. Available in dbt Cloud's "Latest" release track and dbt v1.9 and later.

Author: confirm the directory/paths before committing 
```suggestion
Updated syntax for defining snapshot configurations in YAML files. Found in `whatever_name.yml` or in a snapshots or models directory you prefer. Available in dbt Cloud's "Latest" release track and dbt v1.9 and later.
```

Author: most important thing to specify up top -- benefits of yaml is that the snapshots are environment aware - dont have to specify schema/database

#### Use cases

> - More performant and easier to manage.

Author: ```suggestion
```

Snapshots can be configured in two main ways: 
- Using [snapshot-specific configurations](#snapshot-specific-configurations)
- Or using [general configurations](#general-configuration) 

These configurations allow you to control how dbt detects changes in your data and where snapshots are stored. Both types of configurations can coexist in your project in the same `config` block (or from your `dbt_project.yml` file or `properties.yaml` file). You can also configure snapshots using [strategies](#snapshot-strategies), which define how dbt knows if a row has changed.

### Snapshot specific configurations
Snapshot-specific configurations are applicable to only one dbt resource type rather than multiple resource types. You can define these settings within the resource’s file using the `{{ config() }}` macro (as well as in the project file (`dbt_project.yml`) or a property file (`models/properties.yml` for models, similarly for other resources)).

<File name='snapshots/orders_snapshot.sql'>

```sql
{ % snapshot orders_snapshot %}

{{ config(

>     [target_schema](/reference/resource-configs/target_schema)="<string>",

Author: confirm if `schema` and `database` applies to legacy sql snapshot:

Developers can still set a custom location with [schema](https://docs.getdbt.com/reference/resource-configs/schema) and [database](https://docs.getdbt.com/reference/resource-configs/database) configs, consistent with other resource types.

Author: confirmed

Author: ![Screenshot 2025-03-03 at 13 07 06](https://github.com/user-attachments/assets/7691b39d-acfb-4317-bee1-7e3f7c552106)

    }}

    select * from {{ source('jaffle_shop', 'orders') }}

{% endsnapshot %}
```
</File>

#### Examples
<Expandable alt_header="Check a list of columns for changes">

<File name='snapshots/check_example.sql'>

```sql
{% snapshot orders_snapshot_check %}

    {{
        config(
          strategy='check',
          unique_key='id',
          check_cols=['status', 'is_cancelled'],
        )
    }}

    select * from {{ source('jaffle_shop', 'orders') }}

{% endsnapshot %}
```
</File>
</Expandable>

<Expandable alt_header="Check all columns for changes">

<File name='snapshots/check_example.sql'>

```sql
{% snapshot orders_snapshot_check %}

    {{
        config(
          strategy='check',
          unique_key='id',
          check_cols='all',
        )
    }}

    select * from {{ source('jaffle_shop', 'orders') }}

{% endsnapshot %}
```
</File>
</Expandable>
</TabItem>
</Tabs>

## Configure snapshots

In dbt versions 1.8 and earlier, snapshots are `select` statements, defined within a snapshot block in a `.sql` file (typically in your `snapshots` directory or any other directory). You'll also need to configure your snapshot to tell dbt how to detect record changes.

> The following table outlines the configurations available for snapshots in versions 1.8 and earlier:

Author: ```suggestion
The following table outlines the configurations available for snapshots:
```

| Config | Description | Required? | Example |
| ------ | ----------- | --------- | ------- |
| [target_database](/reference/resource-configs/target_database) | The database that dbt should render the snapshot table into | No | analytics |

> | [target_schema](/reference/resource-configs/target_schema) | The schema that dbt should render the snapshot table into | Yes | snapshots |

Author: test out

| [strategy](/reference/resource-configs/strategy) | The snapshot strategy to use. One of `timestamp` or `check` | Yes | timestamp |
| [unique_key](/reference/resource-configs/unique_key) | A <Term id="primary-key" /> column or expression for the record | Yes | id |
| [check_cols](/reference/resource-configs/check_cols) | If using the `check` strategy, then the columns to check | Only if using the `check` strategy | ["status"] |
| [updated_at](/reference/resource-configs/updated_at) | If using the `timestamp` strategy, the timestamp column to compare | Only if using the `timestamp` strategy | updated_at |
| [invalidate_hard_deletes](/reference/resource-configs/invalidate_hard_deletes) | Find hard deleted records in source, and set `dbt_valid_to` current time if no longer exists | No | True |

> - A number of other configurations are also supported (e.g. `tags` and `post-hook`), check out the full list [here](/reference/snapshot-configs).

Author: ```suggestion
- A number of other configurations are also supported (like, `tags` and `post-hook`), check out the full list [here](/reference/snapshot-configs).
```

- Snapshots can be configured from both your `dbt_project.yml` file and a `config` block, check out the [configuration docs](/reference/snapshot-configs) for more information.
- Note: BigQuery users can use `target_project` and `target_dataset` as aliases for `target_database` and `target_schema`, respectively.

## Add snapshot to a project

To add a snapshot to your project:

> 1. Create a file in your `snapshots` directory with a `.sql` file extension, e.g. `snapshots/orders.sql`

Author: ```suggestion
1. Create a file in your `snapshots` directory with a `.sql` file extension. For example,`snapshots/orders.sql`
```

  {{
      config(
        target_schema='snapshots',
        unique_key='id',

        strategy='timestamp',
        updated_at='updated_at_for_snapshot'
      )
  }}

  select
      *,
      coalesce(updated_at, created_at) as updated_at_for_snapshot

  from {{ source('jaffle_shop', 'orders') }}

  {% endsnapshot %}
  ```
  </File>
</Expandable>

<Expandable alt_header="Using the unique_key parameter">

The `unique_key` is a column name or expression that is unique for the inputs of a snapshot. dbt uses [`unique_key`](/reference/resource-configs/unique_key) to match records between a result set and an existing snapshot, so that changes can be captured correctly.

<File name='snapshots/orders.sql'>

```sql
{{ config(
  unique_key="column_name"
) }}
```
</File>

#### Examples

- Using an `id` column as a unique key

  <File name='snapshots/orders.sql'>

  ```sql
  {{
      config(
        unique_key="id"
      )
  }}
  ```
  </File>

  You can also write this in YAML. This might be a good idea if multiple snapshots share the same `unique_key` (though we prefer to apply this configuration in a config block, as above).

- #### Using a combination of two columns as a unique key

>   This configuration accepts a valid column expression. As such, you can concatenate two columns together as a unique key if required. It's a good idea to use a separator (e.g. '-') to ensure uniqueness.

Author: ```suggestion
  This configuration accepts a valid column expression. As such, you can concatenate two columns together as a unique key if required. It's a good idea to use a separator (like, '-') to ensure uniqueness.
```

  <File name='snapshots/transaction_items_snapshot.sql'>

  ```sql
  {% snapshot transaction_items_snapshot %}

      {{
          config(
            unique_key="transaction_id||'-'||line_item_id",
            ...
          )
      }}

  select
      transaction_id||'-'||line_item_id as id,
      *
  from {{ source('erp', 'transactions') }}

  {% endsnapshot %}
  ```

  </File>

  Though, it's probably a better idea to construct this column in your query and use that as the `unique_key`:

    <File name='snapshots/transaction_items_snapshot.sql'>

    ```sql
    {% snapshot transaction_items_snapshot %}

        {{
            config(
              unique_key="id",
              ...
            )
        }}

    select
        transaction_id || '-' || line_item_id as id,
        *
    from {{ source('erp', 'transactions') }}

    {% endsnapshot %}
    ```
    </File>
</Expandable>

> ## Migrate legacy snapshot configs

Author: remove this section. and could be more of a devblog. ' you dont need to, but if you want to...here's how to do it.'
